src/test/java/baseball/UniqueNumberStringTest.java

```diff
@@ -0,0 +1,111 @@
+        // given
+        String[][] cases = {{"123", "321"}, {"123", "123"}, {"123", "231"}, {"123", "456"}};
+        int[][] answers = {{2, 1}, {0, 3}, {3, 0}, {0, 0}};
+
+        for (int i = 0; i < cases.length; i++) {
+            // when
+            UniqueNumberString numberString1 = UniqueNumberString.create(cases[i][0], RuleSet.BASIC);
+            UniqueNumberString numberString2 = UniqueNumberString.create(cases[i][1], RuleSet.BASIC);
+
+            MatchResult result = UniqueNumberString.match(numberString1, numberString2);
+
+            // then
+            assertEquals(result.getBall(), answers[i][0],
+                    "case " + cases[i][0] + ", " + cases[i][1] + "에서 ball 불일치");
+            assertEquals(result.getStrike(), answers[i][1],
+                    "case " + cases[i][0] + ", " + cases[i][1] + "에서 strike 불일치");
+        }
+    }
+
+    @Test
+    void create_Test() {
+        // given
+        String[] cases = {"123", "456", "789", "112", "121", "211", "111", "", "1", "11", "1111", "cow", "24L"};
+        boolean[] answers = {true, true, true, false, false, false, false, false, false, false, false, false, false};
+
+        for (int i = 0; i < cases.length; i++) {
+            // when
+            String value = cases[i];
+
+            // then
+            if (answers[i]) {
+                assertDoesNotThrow(() -> UniqueNumberString.create(value, RuleSet.BASIC),
+                        "올바른 값을 생성하지 못했습니다. : " + cases[i]);
+            } else {
+                assertThrows(IllegalArgumentException.class,
+                        () -> UniqueNumberString.create(value, RuleSet.BASIC),
+                        "옳지 않은 값이 생성되었습니다. : " + cases[i]);
+            }
+        }
+    }
+
+    @Test
+    void CompareResult_toString_Test() {
+        // given
+        String[][] cases = {{"123", "321"}, {"123", "123"}, {"123", "231"}, {"123", "456"}};
+        String[] answers = {"2볼 1스트라이크", "3스트라이크", "3볼", "낫싱"};
+
+        for (int i = 0; i < cases.length; i++) {
+            // when
+            UniqueNumberString numberString1 = UniqueNumberString.create(cases[i][0], RuleSet.BASIC);
+            UniqueNumberString numberString2 = UniqueNumberString.create(cases[i][1], RuleSet.BASIC);
+
+            MatchResult result = UniqueNumberString.match(numberString1, numberString2);
+
+            // then
+            assertEquals(result.toString(), answers[i]);
+        }
+    }
+
+    @Test
+    void createRandom_Test() {
+        // given
+        int equalCount = 0;
+        int equalLimit = 5;
+
+        // when
+        for (int i = 0; i < 100; i++) {
+            UniqueNumberString numberString1 = UniqueNumberString.createRandom(RuleSet.BASIC);
+            UniqueNumberString numberString2 = UniqueNumberString.createRandom(RuleSet.BASIC);
+
+            if (numberString1.equals(numberString2)) {
+                equalCount++;
+            }
+        }
+
+        // then
+        assertTrue(equalCount < equalLimit);
+    }
```

Comment:
테스트코드에 반복문이 들어가게 되면 테스트를 읽는 데 많은 시간이 걸린다고 생각해요! 그래서 아래의 방법을 추천합니다 :)
```suggestion
    @RepeatedTest(100)
    @Test
    void createRandom_Test() {
        // given
        int equalCount = 0;
        int equalLimit = 5;

        // when
        UniqueNumberString numberString = UniqueNumberString.createRandom(RuleSet.BASIC);
        //then
        // 세자리 중복된 숫자가 없는지 검증 테스트코드
        }
        assertTrue(equalCount < equalLimit);
    }
```
RepeatedTest 애노테이션을 활용하면 같은 테스트를 지정한 수 만큼 반복합니다. 랜덤한 값이 정해진 규칙에 맞게 생성하는지 검증한다면 정상적으로 생성하는지 검증할 수 있을 것 같아요 :)

Reply:
다양한 어노테이션이 있네요. 좋은 정보 감사합니다😊

그런데 이 상황에서는 사용하기 애매해보입니다. 랜덤값을 2개 연속생성 했을때 두개가 같을확률이 5%넘어가면 정말 큰 문제가 있다 라는 의미로 작성한 테스트입니다. 역치값이 limit에 적혀있습니다.
equalCount를 세어야 하는데 검색해보니 스코프가 유지되는거같지는 않더군요. 이를 위해 JUnit을 통한 추가적인 조치를 하기보다는 간단한 테스트는 간단하게 놔둬도 괜찮다고 생각합니다.

혹시 이 테스트에 한해서 읽는데 불편한 점이 있으셨나요?

Reply:
개인적으로 테스트 코드를 보고 느낀 것 중의 하나가 이름이 명확하지 않았다고 생각합니다.
현재 테스트 메서드는 createRandom_Test인데 이 테스트가 <u>_랜덤한 값 생성_</u>과 <u>_역치값을 확인_</u>  중에서 무엇이 맞는지는 코드를 보고 판단해야 한다고 생각합니다.

만약 단순하게 아래와 같이 작성했다면 이해하기 쉬웠으리라 생각합니다.
```java
void 랜덤한_생성값은_역치값을_넘기지_않는다(){
```

두 번째로 랜덤값을 2개 연속 생성 했을 경우 두 값이 5%가 넘어간다는 건 테스트가 언제든 <b>실패</b>할 가능성이 있다고 생각합니다.
즉 현재는 테스트가 통과할지라도 랜덤값을 2개 연속 생성해서 5번 정도 같아진다는 것은 그렇게 낮은 확률은 아니라고 생각합니다.

단위 테스트는 언제 실행하든 같은 결과를 도출해야 하는 특징을 가지고 있으므로 확률에 기대는 것은 단위 테스트의 성격과는 거리가 멀다고 생각합니다.

c#이지만 정말 잘 정리된 문서라고 생각합니다:)
https://learn.microsoft.com/ko-kr/dotnet/core/testing/unit-testing-best-practices

![스크린샷 2023-10-28 오후 5 51 06](https://github.com/woowacourse-precourse/java-baseball-6/assets/73726591/db0cd8ee-bf91-436d-9833-89afae82ae34)

Reply:
코드가 DDD 기반이라 테스트를 보기 전에 상대방이 도메인을 어떻게 바라봤는지를 먼저 보는게 맞다고 생각해요. 음, 관련내용은 말씀드리기 조심스럽네요. 저도 TDD에서 어떤 이야기를 하는지 알고있습니다. 다만 10줄 남짓의 코드에서 오해가 생길거라 예상하지는 못했어요.

두번째 말씀은 동의하기 어렵습니다. 먼저 난수 분포의 적절성에 대한 테스트는 필요하다고 생각해요. 난수생성이 아닌 객체 생성에서 문제가 생길수도 있으니까요.

보내주신 문서는 잘 읽었습니다. 하지만 저건 "좋은" 단위테스트의 특징입니다. 좋지 않다고 해서 필요하지 않다는건 아니라고 생각해요. 해당 테스트의 레벨도 단위테스트 레벨이 적절하다고 생각하구요.

---

테스트 관점에서는 말씀하셨던 실패가능성과 난수테스트 방법에 대해 같이 고민해보고 싶어서 리뷰를 열었어요.

지금 드는 생각은 무조건 pass 시키되, 로그나 시각자료를 출력하여 최종적으로 사람이 판단하게하는 방식이 좋아보입니다. 실제로 성능테스트도 크게 엄격한 항목이 아니라면 프로파일링을 통한 보고서 출력을 하는 경우가 많더군요.

번외로 테스트 실패확률은 극악입니다. 현재조건에서 2개의 랜덤값을 생성했을때 일치할 확률이 0.13%에요. 5번까지 허용하면 이항분포로 계산해야 합니다. 계산 해보면 0.000001% 미만이에요. 이걸 생각하지 않고 작성하진 않았어요. 절대 일어나지 않아야 할 일이 일어나는 것 만은 체크하자는 차원에서 작성했죠. 물론 적절한 방법은 아니었어요. [엔트로피나 확률분포](https://medium.com/unitychain/provable-randomness-how-to-test-rngs-55ac6726c5a3)를 사용하는게 훨씬 목적에 맞았습니다.

어찌되었든 테스트 명이든 주석이든 더 구체적으로 작성했어야 했습니다. 큰 반성을 하고 있습니다.

---

귀한 시간 내주셔서 감사합니다. 혹시라도 기분나쁘셨던 부분이 있다면 사과드립니다.

1주차 고생하셨고, 2주차도 화이팅입니다.

Reply:
아닙니다. 전혀 기분 나쁘지는 않았습니다 :)
저는 오히려 생각하지 못했던 테스트 케이스를 구성하셨단 사실에 감탄했습니다. 👍 
만약 테스트를 위해 반복적으로 생성한 숫자가 요구사항에 맞게 출력된다고 해서 그 숫자가 정말로 랜덤한 숫자로 구성되었다는 사실을 검증하지 못할 수  있겠다는 사실을 효재님 코드를 보고 알게 되었습니다 :)

오히려 테스트 관련해서 이야기해볼 수 있는 좋은 주제였다고 생각합니다. 덕분에 저도 한 단계 성장할 수 있었네요!
아마 이런 부분이 코드리뷰의 장점이 아닐까 생각해요ㅎㅎ

Reply:
기분이 안나쁘셨다니 마음이 놓이네요😊 감사합니다. 좋은 밤 보내세요!